lib/src/concat.dart

```diff
@@ -17,11 +20,66 @@ class _Concat<T> implements StreamTransformer<T, T> {
 
   @override
   Stream<T> bind(Stream<T> first) {
-    var controller = new StreamController<T>();
-    controller
-        .addStream(first)
-        .then((_) => controller.addStream(_next))
-        .then((_) => controller.close());
+    var controller = first.isBroadcast
+        ? new StreamController<T>.broadcast(sync: true)
+        : new StreamController<T>(sync: true);
+
+    StreamSubscription subscription;
+    var currentStream = first;
+    var firstDone = false;
+    var secondDone = false;
+
+    Function currentDoneHandler;
+
+    listen() {
+      subscription = currentStream.listen(controller.add,
+          onError: controller.addError, onDone: () => currentDoneHandler());
+    }
+
+    onSecondDone() {
+      secondDone = true;
+      controller.close();
+    }
+
+    onFirstDone() {
+      firstDone = true;
+      currentStream = _next;
+      currentDoneHandler = onSecondDone;
+      listen();
+    }
+
+    currentDoneHandler = onFirstDone;
+
+    controller.onListen = () {
+      if (firstDone &&
```

Comment:
The code paths around re-listening should only get hit if `first` is a broadcast stream, otherwise the controller would prevent us from ever getting here.

Our first listener can't hit this since `firstDone` won't ever be true. Subsequent listeners can only hit if `first` was a broadcast stream. We're not still listening to `first` since `firstDone` is true, which means we are (were) listening to `_next` so it's the only stream that `subscription` could be on.